migrator/migrations/m_181_to_m_182_remove_default_scope_manager_role/migration_impl.go

```diff
@@ -0,0 +1,327 @@
+		}
+	} else {
+		canRemovePermissionSet = false
+	}
+	// Check for permission set references
+	usesScopeManagerPermissionSet, usePermissionSetCheckErr := isScopeManagerPermissionSetReferenced(ctx, roleStorage)
+	if usePermissionSetCheckErr != nil {
+		canRemovePermissionSet = false
+		err = multierror.Append(err, usePermissionSetCheckErr)
+	}
+	if usesScopeManagerPermissionSet {
+		// The Role -> PermissionSet reference is using the PermissionSet ID which will be left untouched.
+		// Therefore, there is no need to update the PermissionSet reference in the roles.
+		// The permission set Name and Description will be updated.
+		updatePermissionSetErr := updateScopeManagerPermissionSet(ctx, permissionSetStorage)
+		if updatePermissionSetErr != nil {
+			err = multierror.Append(err, updatePermissionSetErr)
+		}
+	} else {
+		if canRemovePermissionSet {
+			deletePermissionSetErr := removeDefaultScopeManagerPermissionSet(ctx, permissionSetStorage)
+			if deletePermissionSetErr != nil {
+				err = multierror.Append(err, deletePermissionSetErr)
+			}
+		}
+	}
+	return err.ErrorOrNil()
+}
+
+func apiTokenHasScopeManagerRole(obj *storage.TokenMetadata) bool {
+	tokenRoles := obj.GetRoles()
+	for _, roleName := range tokenRoles {
+		if roleName == ScopeManagerRoleName {
+			return true
+		}
+	}
+	return false
+}
+
+func groupHasScopeManagerRole(obj *storage.Group) bool {
+	return obj.GetRoleName() == ScopeManagerRoleName
+}
+
+func isScopeManagerRoleReferenced(ctx context.Context, apiTokenStorage apiTokenStore.Store, groupStorage groupStore.Store) (bool, error) {
+	roleReferenceFound := false
+	var err *multierror.Error
+	tokenWalkErr := apiTokenStorage.Walk(ctx, func(obj *storage.TokenMetadata) error {
+		if roleReferenceFound {
+			return nil
+		}
+		hasScopeManagerRole := apiTokenHasScopeManagerRole(obj)
+		if hasScopeManagerRole {
+			roleReferenceFound = true
+		}
+		return nil
+	})
+	if tokenWalkErr != nil {
+		err = multierror.Append(err, tokenWalkErr)
+	}
+	if roleReferenceFound {
+		return true, nil
+	}
+	groupWalkErr := groupStorage.Walk(ctx, func(obj *storage.Group) error {
+		if roleReferenceFound {
+			return nil
+		}
+		if groupHasScopeManagerRole(obj) {
+			roleReferenceFound = true
+		}
+		return nil
+	})
+	if groupWalkErr != nil {
+		err = multierror.Append(err, groupWalkErr)
+	}
+	multiErr := err.ErrorOrNil()
+	if multiErr != nil {
+		return false, multiErr
+	}
+	return roleReferenceFound, nil
+}
+
+func addDeprecatedScopeManagerRole(ctx context.Context, roleStorage roleStore.Store) error {
+	oldRole, oldRoleFound, err := roleStorage.Get(ctx, ScopeManagerRoleName)
+	if err != nil {
+		return err
+	}
+	if !oldRoleFound {
+		return errors.Wrap(errNotFound, "looking up role")
+	}
+	newRole := oldRole.Clone()
+	newRole.Name = deprecatedPrefix + ScopeManagerRoleName
+	newRole.Traits = &storage.Traits{
+		// Keep default MutabilityMode : Traits_ALLOW_MUTATE
+		// Keep default Visibility : Traits_VISIBLE
+		Origin: storage.Traits_IMPERATIVE,
```

Comment:
I assume the reasoning behind marking this as IMPERATIVE (which you don't need to explicitly set, it's the default value), is because we expect customers to delete the role afterwards manually?

Reply:
What customers do with the `Scope Manager` role and permission set after the migration is their decision.
The point here is to reflect the fact that the role and permission set are not inserted/maintained by the product as default anymore.